src/workerd/api/streams/standard.c++

```diff
@@ -2080,6 +2103,184 @@ ReadableStreamJsController::getController() {
+    state.template init<StreamStates::Errored>(js.v8Ref(reason));
+  }
+
+  jsg::Promise<kj::Array<byte>> allBytes(jsg::Lock& js) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
+      auto out = kj::heapArray<byte>(runningTotal);
+      copyInto(out, kj::mv(partPtrs));
+      return kj::mv(out);
+    });
+  }
+
+  jsg::Promise<kj::String> allText(jsg::Lock& js) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
+      auto out = kj::heapArray<char>(runningTotal + 1);
+      copyInto(out.slice(0, out.size() - 1).asBytes(), kj::mv(partPtrs));
+      out.back() = '\0';
+      return kj::String(kj::mv(out));
+    });
+  }
+private:
+  kj::OneOf<StreamStates::Closed, StreamStates::Errored, Readable> state;
+  uint64_t limit;
+  kj::Vector<jsg::BackingStore> parts;
+  uint64_t runningTotal = 0;
+
+  jsg::Promise<PartList> loop(jsg::Lock& js) {
+    KJ_SWITCH_ONEOF(state) {
+      KJ_CASE_ONEOF(closed, StreamStates::Closed) {
+        return js.resolvedPromise(KJ_MAP(p, parts) { return p.asArrayPtr(); });
+      }
+      KJ_CASE_ONEOF(errored, StreamStates::Errored) {
+        return js.template rejectedPromise<PartList>(errored.getHandle(js));
+      }
+      KJ_CASE_ONEOF(readable, Readable) {
+        const auto read = [&](auto& js) {
+          if constexpr (kj::isSameType<T, ByteReadable>()) {
+            return readable->read(js, nullptr);
+          } else {
+            return readable->read(js);
+          }
+        };
+
+        return read(js).then(js, [this](auto& js, ReadResult result) {
+          if (result.done) {
+            KJ_ASSERT(result.value == nullptr);
+            doClose();
+            return loop(js);
+          }
+
+          // If we're not done, the result value must be interpretable as
+          // bytes for the read to make any sense.
+          auto handle = KJ_ASSERT_NONNULL(result.value).getHandle(js);
+          if (!handle->IsArrayBufferView() && !handle->IsArrayBuffer()) {
+            auto error = js.v8TypeError("This ReadableStream did not return bytes.");
+            doError(js, error);
+            return loop(js);
+          }
+
+          jsg::BufferSource bufferSource(js, handle);
+          jsg::BackingStore backing = bufferSource.detach(js);
+          runningTotal += backing.size();
+          if (runningTotal >= limit) {
+            auto error = js.v8TypeError("Memory limit exceeded before EOF.");
+            doError(js, error);
+            return loop(js);
+          }
+
+          parts.add(kj::mv(backing));
+          return loop(js);
+        }, [this](auto& js, jsg::Value exception) {
+          doError(js, exception.getHandle(js));
+          return loop(js);
+        });
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+
+  void copyInto(kj::ArrayPtr<byte> out, PartList in) {
+    size_t pos = 0;
+    for (auto& part: in) {
+      KJ_ASSERT(part.size() <= out.size() - pos);
+      memcpy(out.begin() + pos, part.begin(), part.size());
```

Comment:
I know this is safe, but let's use `std::copy` for consistency with the rest of our new code.